### ISSUE TITLE
Rendering Layer Not Differentiable. Cannot train with wild images. 

### ISSUE THREAD

**Issue by chaoshiedwin** [tranluan/Nonlinear_Face_3DMM#59]
Hi @tranluan Thank you for your work. I am trying to fine-tune my model with wild images.  But in your code, the z-buffer kernel is marked as NotDifferentiable, which means the reconstruction loss cannot be back propagated to the network. Hence wild images cannot be used for network training.  This is inconsistent with what is claimed in the paper. Can you provide the gradient definition?

**Comment by chaoshiedwin** [comment]
Sorry, I didn't understand it correctly. The z-buffer does not need to be differentiable, z-buffer output is used to calculate the sampling position from texture map.

**Comment by HOMGH** [comment]
> Sorry, I didn't understand it correctly. The z-buffer does not need to be differentiable, z-buffer output is used to calculate the sampling position from texture map.

@chaoshiedwin  Hi, Could you train the model successfully?  I trained the model on the 1/4 of training set and it seems it doesn't converge!

**Comment by huyanfei-cqupt** [comment]
> Hi @tranluan Thank you for your work. I am trying to fine-tune my model with wild images. But in your code, the z-buffer kernel is marked as NotDifferentiable, which means the reconstruction loss cannot be back propagated to the network. Hence wild images cannot be used for network training. This is inconsistent with what is claimed in the paper. Can you provide the gradient definition?

hi,@chaoshiedwin Could you train the model successfully?
